### PR TITLE
GH Issue #49

### PR DESCRIPTION
**Pull request by jpeiffer**

### CODE REVIEW

**Comment by jpeiffer** [comment]
Implemented using a pseudo-enum mechanism.  Opted not to do a true enum since there's no way to translate an enum to a string name (vote for: https://github.com/dart-lang/sdk/issues/21712).

Please let me know if there are any changes you'd like me to make before merging.

**Comment by kevmoo** [comment]
@jonasfj ? CC @mit-mit

**Comment by kevmoo** [comment]
er @sortie –  sorry @jonasfj

**Comment by sortie** [comment]
@kevmoo The ownership of pub has been transferred to @jonasfj. I don't know anything about this repository.

**Comment by jpeiffer** [comment]
I see that the Firefox test failed on Travis.  When I ran `pub run test` locally everything passed.  Is there a guide for running the full suite locally so I can diagnose and fix the failure?

**Comment by jonasfj** [comment]
@jpeiffer, I think you just do `pub run test -p firefox` to run on firefox... same for chrome...

Note, the travis file says `pub run test -p firefox -j 1`.

**Comment by jpeiffer** [comment]
So, I have to ask, do these false negative a bit?

I did multiple test runs locally.  The results from the first four are:
1) Everything passed except:
```
00:17 +0 -1: compiling test/crypto/asn1_test.dart [E]                                                                                                        
  Failed to load "test/crypto/asn1_test.dart":
  Failed to run Firefox: No such file or directory
    Command: /Applications/Firefox.app/Contents/MacOS/firefox-bin --profile /private/var/folders/b4/9_t84stn14l7c5f8_8dr4sd00000gn/T/dart_test_VjtwHb http://localhost:59062/zhpfTJPZb1xlNa7lzB3Ezd66MGfR43aX/packages/test/src/runner/browser/static/index.html?managerUrl=ws%3A%2F%2Flocalhost%3A59062%2FzhpfTJPZb1xlNa7lzB3Ezd66MGfR43aX%2F0&debug=false --no-remote.
```

2) Everything passed except:
```
00:16 +0 -1: compiling test/crypto/pem_test.dart [E]                                                                                                         
  Failed to load "test/crypto/pem_test.dart":
  Failed to run Firefox: No such file or directory
    Command: /Applications/Firefox.app/Contents/MacOS/firefox-bin --profile /private/var/folders/b4/9_t84stn14l7c5f8_8dr4sd00000gn/T/dart_test_2gfEE7 http://localhost:59141/sEEIU7aPe2iTo3sQhcU5PnkWqOGjbw25/packages/test/src/runner/browser/static/index.html?managerUrl=ws%3A%2F%2Flocalhost%3A59141%2FsEEIU7aPe2iTo3sQhcU5PnkWqOGjbw25%2F0&debug=false --no-remote.
```

3) Everything passed except...
```
00:16 +0 -1: compiling test/crypto/asn1_test.dart [E]                                                                                                        
  Failed to load "test/crypto/asn1_test.dart":
  Failed to run Firefox: No such file or directory
    Command: /Applications/Firefox.app/Contents/MacOS/firefox-bin --profile /private/var/folders/b4/9_t84stn14l7c5f8_8dr4sd00000gn/T/dart_test_HFBnGT http://localhost:59216/FghiL97%2B6uW8Pdg%2BwVqpIDSk1wadwsOE/packages/test/src/runner/browser/static/index.html?managerUrl=ws%3A%2F%2Flocalhost%3A59216%2FFghiL97%252B6uW8Pdg%252BwVqpIDSk1wadwsOE%2F0&debug=false --no-remote.
```

4) Everything passed except...
```
00:08 +0 -1: compiling test/crypto/rsa_test.dart [E]                                                                                                         
  Failed to load "test/crypto/rsa_test.dart":
  Failed to run Firefox: No such file or directory
    Command: /Applications/Firefox.app/Contents/MacOS/firefox-bin --profile /private/var/folders/b4/9_t84stn14l7c5f8_8dr4sd00000gn/T/dart_test_8yK7UE http://localhost:59289/dV2loqr6WBx%2B8GqJ%2Fj1QAXv1tZsBjjHf/packages/test/src/runner/browser/static/index.html?managerUrl=ws%3A%2F%2Flocalhost%3A59289%2FdV2loqr6WBx%252B8GqJ%252Fj1QAXv1tZsBjjHf%2F0&debug=false --no-remote.
```

I was able to get the test that Travis failed to fail by re-running several more times, but I don't think the issue is in the code or the test code at this point.  All but one time I received:
```
02:19 +0 -1: compiling test/oauth2_flows/implicit/gapi_auth_implicit_idtoken_test.dart                                                                       
Compiled 8,631,118 characters Dart to 1,063,120 characters JavaScript in 7.81 seconds
Dart file (/private/var/folders/b4/9_t84stn14l7c5f8_8dr4sd00000gn/T/dart_test_QchAmJ/runInBrowser.dart) compiled to JavaScript: /private/var/folders/b4/9_t84stn14l7c5f8_8dr4sd00000gn/T/dart_test_cQez7s/test_0vljEI/gapi_auth_implicit_idtoken_test.dart.browser_test.dart.js
```

**Comment by jonasfj** [comment]
@jpeiffer, if you read the error message it says:
>   Failed to run Firefox: No such file or directory

Which prompts me to ask the obvious, do you have firefox installed? :smile: 

I had no problems running `pub run test -p firefox` on my linux desktop.
You could try to run it in docker under `xfvb-run pub run test -p firefox`, but installing firefox and xvfb isn't trivial and it could be hard to do. First check if you have firefox installed on your machine, and if so, is it in `/Applications/Firefox.app/Contents/MacOS/firefox-bin` ?

File: lib/auth.dart [hunk 1]
```diff
@@ -285,4 +285,34 @@ Future<AccessCredentials> refreshCredentials(
...
+  @override
+  bool operator ==(other) {
+    return (other is ResponseType) && other.value == value;
+  }
+
+  @override
+  int get hashCode {
+    return ('ResponseType.' + value).hashCode;
+  }
+}
```
Comment (jonasfj) [inline lib/auth.dart]: How about we use an enum like:
```dart
enum ReponseType {
  code,
  idToken,
  permission,
}


String responseTypeToString(ResponseType t) {
  switch (t) {
    case RepsonseType.code:
      return 'code';
    case RepsonseType.idToken:
      return 'id_token';
    case RepsonseType.permission:
      return 'permission';
  }
  throw ArgumentError('unknown ResponseType: $t');
}
```
If we move `responseTypeToString` to another file, we can probably make it private too :)

**Comment by jpeiffer** [comment]
Ha!  I just had my laptop replaced and just realized FF wasn't installed.  It threw me that a different test failed each time.  I completely overlooked that message.  My bad.  ;)

Trying out your requested changes now.

File: lib/auth.dart (hunk 1 above)
Comment (jpeiffer) [reply lib/auth.dart]: Done.

**Comment by jpeiffer** [comment]
Thanks for your patience with a newbie on Dart Web.  I found the issue with the test and made the changes you requested.  

Please let me know if there are any further changes you'd like.

File: lib/src/oauth2_flows/implicit.dart [hunk 2]
```diff
@@ -162,3 +180,30 @@ class LoginResult {
...
+      result = 'id_token';
+      break;
+
+    case ResponseType.permission:
+      result = 'permission';
+      break;
+
+    case ResponseType.token:
+      result = 'token';
+      break;
```
Comment (jonasfj) [inline lib/src/oauth2_flows/implicit.dart]: Nit... but you could make this a bit simpler with:
```dart
/// Convert [responseType] to string value expected by `gapi.auth.authorize`.
String _responseTypeToString(ResponseType responseType) {
  switch(responseType) {
    case ResponseType.code:
      return 'code';
    case ResponseType.idToken:
      return 'id_token';
    case ResponseType.token:
      return 'token';
  }
  throw ArgumentError('unknown response type: $responseType');
}
```

File: lib/auth.dart [hunk 3]
```diff
@@ -285,4 +285,13 @@ Future<AccessCredentials> refreshCredentials(
       idToken: idToken);
 }
 
+/// Available response types that can be requested when using the implicit
+/// browser login flow.
+enum ResponseType {
+  code,
```
Comment (jonasfj) [inline lib/auth.dart]: If you want to be really nice, add a short doc-comment to each of these entrie s :)

Just a one-liner would go far...

**Comment by jonasfj** [comment]
Note. this is just nits, I think we're good to land this.. if you're quick to fix the nits I can merge and release tomorrow maybe...

**Comment by jpeiffer** [comment]
Documentation changes done

**Pull request merged**